### PR TITLE
Optimize amp for multi-gpu to enable FP16 gradients transfer across gpus

### PR DESCRIPTION
**Pull request by Jie-Fang**
For the original amp, all the gradients are FP32. Thus in multi-gpus training, FP32 gradients are transfered across gpus which may cause communication overhead. In the amp, we will obtain FP16 gradients of some ops automatically and we can transfer them instead to reduce the overhead.

### CODE REVIEW

File: python/paddle/fluid/contrib/mixed_precision/fp16_utils.py [hunk 1]
```diff
@@ -178,15 +184,15 @@ def rewrite_program(main_prog, amp_lists):
                         # this in_var isn't the output of other op
                         if in_var.op is None:
                             continue
-                        if in_var.op is op:
+                        elif in_var.op is op:
```
Comment (gongweibao) [inline python/paddle/fluid/contrib/mixed_precision/fp16_utils.py:198]: Why elif?

File: python/paddle/fluid/contrib/mixed_precision/fp16_utils.py [hunk 2]
```diff
@@ -130,6 +130,13 @@ def _insert_cast_op(block, op, idx, src_dtype, dest_dtype):
 
 
 def find_true_prev_op(ops, var_name):
+    """
+    Find the true prev op that outputs var_name variable.
+
+    Args:
+        ops (list): A list of ops.
+        var_name (string): Variable name.
+    """
```
Comment (gongweibao) [inline python/paddle/fluid/contrib/mixed_precision/fp16_utils.py:140]: I think we should assert that there should be only one op which generates this var.

**Review by gongweibao — CHANGES_REQUESTED** [review]

File: python/paddle/fluid/contrib/mixed_precision/fp16_utils.py (hunk 2 above)
Comment (gongweibao) [reply python/paddle/fluid/contrib/mixed_precision/fp16_utils.py:140]: Just for safe!

File: python/paddle/fluid/contrib/mixed_precision/fp16_utils.py (hunk 1 above)
Comment (Jie-Fang) [reply python/paddle/fluid/contrib/mixed_precision/fp16_utils.py:198]: It's the same. I think elif may be better

**Comment by gongweibao** [comment]
Can you give the performance comparison of Bert Large and Resnet50 with the old?
We should approve that float16 allreduce can improve communication.

**Comment by Jie-Fang** [comment]
Yes, I think so. But I need ResNet50 multi-gpu version. The script in the models/PaddleCV/images_classification seems not work. It's OK for me to fix the bugs , but it'll takes more time. It's better you have correct script.

File: python/paddle/fluid/contrib/mixed_precision/fp16_utils.py (hunk 2 above)
Comment (Jie-Fang) [reply python/paddle/fluid/contrib/mixed_precision/fp16_utils.py:140]: I'll fix it

**Comment by gongweibao** [comment]
I see.
But I think this pr should test on multi-GPUs to prevent the program bugs. Can you try this?
https://github.com/PaddlePaddle/Fleet/tree/develop/benchmark/collective/resnet

I will test the performance on multi-gpu machines.

**Comment by gongweibao** [comment]
I will test it! Thanks!

**Review by kuke — APPROVED** [review]
LGTM

**Review by gongweibao — APPROVED** [review]
LGTM

**Review by wanghaoshuang — APPROVED** [review]

**Review by hutuxian — APPROVED** [review]

**Pull request merged**